src/app/debugger.js

```diff
@@ -1,39 +1,135 @@
 var remix = require('ethereum-remix');
+var utils = require('./utils');
+var ace = require('brace');
+var Range = ace.acequire('ace/range').Range;
 
-function Debugger (id, executionContextEvent) {
+/**
+ * Manage remix and source highlighting
+ */
+function Debugger (id, editor, compiler, executionContextEvent, switchToFile) {
   this.el = document.querySelector(id);
   this.debugger = new remix.ui.Debugger();
+  this.sourceMappingDecoder = new remix.util.SourceMappingDecoder()
   this.el.appendChild(this.debugger.render());
+  this.editor = editor;
+  this.switchToFile = switchToFile;
+  this.compiler = compiler;
+  this.cache = new Cache()
 
   var self = this;
   executionContextEvent.register('contextChanged', this, function (context) {
-    context = context === 'vm' ? 'VM' : context;
-    context = context === 'injected' ? 'EXTERNAL' : context;
-    context = context === 'web3' ? 'INTERNAL' : context;
     self.switchProvider(context);
   });
-}
 
-Debugger.prototype.debug = function (receipt) {
-  if (this.onDebugRequested) this.onDebugRequested();
+  this.lastCompilationResult = null;
+  this.debugger.register('newTraceLoaded', this, function () {
+    self.cache.clear()
+    self.lastCompilationResult = self.compiler.lastCompilationResult;
+  });
+
+  this.debugger.register('traceUnloaded', this, function () {
+    self.lastCompilationResult = null;
+    self.removeCurrentMarker()
+    self.cache.clear()
+  });
+
+  this.editor.onChangeSetup(function () {
+    if (arguments.length > 0) { // if arguments.length === 0 this is a session change, we don't want to stop debugging in that case
+      self.debugger.unLoad()
+    }
+  });
+
+  // register selected code item, highlight the corresponding source location
+  this.debugger.codeManager.register('changed', this, function (code, address, index) {
+    this.debugger.sourceLocationTracker.getSourceLocation(address, index, self.lastCompilationResult.data.contracts, function (error, rawLocation) {
```

Comment:
This works for now, but in the future, we should not just use "last compilation results". This does not work if you want to debug a past transaction, for example.